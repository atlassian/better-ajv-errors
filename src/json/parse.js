// Custom JSON parser based on https://github.com/douglascrockford/JSON-js/blob/master/json_parse.js
let line;
let column;
let at; // The index of the current character
let ch; // The current character

const META = Symbol('meta');
const VALUE = Symbol('value');

export const getValue = (obj, key) => obj[VALUE][key] || obj[VALUE][VALUE][key];
export const getMeta = obj => obj[META];

const escapee = {
  '"': '"',
  '\\': '\\',
  '/': '/',
  b: '\b',
  f: '\f',
  n: '\n',
  r: '\r',
  t: '\t',
};
let text;

const error = function(m) {
  // Call error when something is wrong.
  throw {
    name: 'SyntaxError',
    message: m,
    at: at,
    text: text,
  };
};

const next = function(c) {
  // If a c parameter is provided, verify that it matches the current character.

  if (c && c !== ch) {
    error("Expected '" + c + "' instead of '" + ch + "'");
  }

  // Get the next character. When there are no more characters,
  // return the empty string.

  ch = text.charAt(at);
  at += 1;
  if (ch === '\n') {
    line += 1;
    column = 0;
  } else {
    column += 1;
  }
  return ch;
};

const number = function() {
  // Parse a number value.

  let value;
  let string = '';

  if (ch === '-') {
    string = '-';
    next('-');
  }
  while (ch >= '0' && ch <= '9') {
    string += ch;
    next();
  }
  if (ch === '.') {
    string += '.';
    while (next() && ch >= '0' && ch <= '9') {
      string += ch;
    }
  }
  if (ch === 'e' || ch === 'E') {
    string += ch;
    next();
    if (ch === '-' || ch === '+') {
      string += ch;
      next();
    }
    while (ch >= '0' && ch <= '9') {
      string += ch;
      next();
    }
  }
  value = +string;
  if (!isFinite(value)) {
    error('Bad number');
  } else {
    return value;
  }
};

const string = function() {
  // Parse a string value.

  let hex;
  let i;
  let value = '';
  let uffff;

  // When parsing for string values, we must look for " and \ characters.

  if (ch === '"') {
    while (next()) {
      if (ch === '"') {
        next();
        return value;
      }
      if (ch === '\\') {
        next();
        if (ch === 'u') {
          uffff = 0;
          for (i = 0; i < 4; i += 1) {
            hex = parseInt(next(), 16);
            if (!isFinite(hex)) {
              break;
            }
            uffff = uffff * 16 + hex;
          }
          value += String.fromCharCode(uffff);
        } else if (typeof escapee[ch] === 'string') {
          value += escapee[ch];
        } else {
          break;
        }
      } else {
        value += ch;
      }
    }
  }
  error('Bad string');
};

const white = function() {
  // Skip whitespace.

  while (ch && ch <= ' ') {
    next();
  }
};

const word = function() {
  // true, false, or null.

  switch (ch) {
    case 't':
      next('t');
      next('r');
      next('u');
      next('e');
      return true;
    case 'f':
      next('f');
      next('a');
      next('l');
      next('s');
      next('e');
      return false;
    case 'n':
      next('n');
      next('u');
      next('l');
      next('l');
      return null;
  }
  error("Unexpected '" + ch + "'");
};

let value; // Place holder for the value function.

const array = function() {
  // Parse an array value.

  const arr = [];
  if (ch === '[') {
    next('[');
    white();
    if (ch === ']') {
      next(']');
      return arr; // empty array
    }
    while (ch) {
      arr.push(value());
      white();
      if (ch === ']') {
        next(']');
        return arr;
      }
      next(',');
      white();
    }
  }
  error('Bad array');
};

const object = function() {
  // Parse an object value.

  let key;
  const obj = {};
  obj[META] = { line, column };
  obj[VALUE] = {};

  if (ch === '{') {
    next('{');
    white();
    if (ch === '}') {
      next('}');
      return obj; // empty object
    }
    while (ch) {
      const c = column;
      key = string();
      white();
      next(':');
      if (Object.hasOwnProperty.call(obj, key)) {
        error("Duplicate key '" + key + "'");
      }
      obj[VALUE][key] = {};
      obj[VALUE][key][META] = { line, column: c };
      obj[VALUE][key][VALUE] = value();
      white();
      if (ch === '}') {
        next('}');
        return obj;
      }
      next(',');
      white();
    }
  }
  error('Bad object');
};

value = function() {
  // Parse a JSON value. It could be an object, an array, a string, a number,
  // or a word.

  white();
  switch (ch) {
    case '{':
      return object();
    case '[':
      return array();
    case '"':
      return string();
    case '-':
      return number();
    default:
      return ch >= '0' && ch <= '9' ? number() : word();
  }
};

// Return the json_parse function. It will have access to all of the above
// functions and variables.
export default source => {
  let result;

  text = source;
  at = 0;
  ch = ' ';
  line = 0;
  column = 0;
  result = value();
  white();
  if (ch) {
    error('Syntax error');
  }
  return result;
};
