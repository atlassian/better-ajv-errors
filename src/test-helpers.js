import { readFileSync } from 'fs';
import { getFixturePath } from 'jest-fixtures';
import Ajv from 'ajv';

export async function getSchemaAndData(name, dirPath) {
  const schemaPath = await getFixturePath(dirPath, name, 'schema.json');
  const schema = JSON.parse(readFileSync(schemaPath, 'utf8'));
  const dataPath = await getFixturePath(dirPath, name, 'data.json');
  const data = JSON.parse(readFileSync(dataPath, 'utf8'));

  return [schema, data];
}

export function evaluateSchema(schema, data) {
  const ajv = new Ajv();
  const fn = ajv.compile(schema);
  fn(data);
  expect(fn.errors.length).toBeGreaterThanOrEqual(0);
  return fn.errors;
}
