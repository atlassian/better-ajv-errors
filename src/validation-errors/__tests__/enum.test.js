import parse from 'json-to-ast';
import { getSchemaAndData } from '../../test-helpers';
import EnumValidationError from '../enum';

describe('Enum', () => {
  describe('when value is an object', () => {
    let schema;
    let data;
    let jsonRaw;
    let jsonAst;

    beforeAll(async () => {
      [schema, data] = await getSchemaAndData('enum', __dirname);
      jsonRaw = JSON.stringify(data, null, 2);
      jsonAst = parse(jsonRaw, { loc: true });
    });

    it.each([
      ['prints correctly for enum prop', true],
      ['prints correctly for enum prop [without colors]', false],
    ])('%s', (_, colorize) => {
      const error = new EnumValidationError(
        {
          keyword: 'enum',
          dataPath: '/id',
          schemaPath: '#/enum',
          params: { allowedValues: ['foo', 'bar'] },
          message: `should be equal to one of the allowed values`,
        },
        { colorize, data, schema, jsonRaw, jsonAst }
      );

      expect(error.print()).toMatchSnapshot();
    });

    it.each([
      ['prints correctly for no levenshtein match', true],
      ['prints correctly for no levenshtein match [without colors]', false],
    ])('%s', (_, colorize) => {
      const error = new EnumValidationError(
        {
          keyword: 'enum',
          dataPath: '/id',
          schemaPath: '#/enum',
          params: { allowedValues: ['one', 'two'] },
          message: `should be equal to one of the allowed values`,
        },
        { colorize, data, schema, jsonRaw, jsonAst }
      );

      expect(error.print()).toMatchSnapshot();
    });

    it.each([
      ['prints correctly for empty value', true],
      ['prints correctly for empty value [without colors]', false],
    ])('%s', (_, colorize) => {
      const error = new EnumValidationError(
        {
          keyword: 'enum',
          dataPath: '/id',
          schemaPath: '#/enum',
          params: { allowedValues: ['foo', 'bar'] },
          message: `should be equal to one of the allowed values`,
        },
        { colorize, data, schema, jsonRaw, jsonAst }
      );

      expect(error.print(schema, { id: '' })).toMatchSnapshot();
    });
  });

  describe('when value is a primitive', () => {
    let schema;
    let data;
    let jsonRaw;
    let jsonAst;

    beforeAll(async () => {
      [schema, data] = await getSchemaAndData('enum-string', __dirname);
      jsonRaw = JSON.stringify(data, null, 2);
      jsonAst = parse(jsonRaw, { loc: true });
    });

    it.each([
      ['prints correctly for enum prop', true],
      ['prints correctly for enum prop [without colors]', false],
    ])('%s', (_, colorize) => {
      const error = new EnumValidationError(
        {
          keyword: 'enum',
          dataPath: '',
          schemaPath: '#/enum',
          params: {
            allowedValues: ['foo', 'bar'],
          },
          message: 'should be equal to one of the allowed values',
        },
        { colorize, data, schema, jsonRaw, jsonAst }
      );

      expect(error.print()).toMatchSnapshot();
    });

    it.each([
      ['prints correctly for no levenshtein match', true],
      ['prints correctly for no levenshtein match [without colors]', false],
    ])('%s', (_, colorize) => {
      const error = new EnumValidationError(
        {
          keyword: 'enum',
          dataPath: '',
          schemaPath: '#/enum',
          params: {
            allowedValues: ['one', 'two'],
          },
          message: 'should be equal to one of the allowed values',
        },
        { colorize, data, schema, jsonRaw, jsonAst }
      );

      expect(error.print()).toMatchSnapshot();
    });

    it.each([
      ['prints correctly for empty value', true],
      ['prints correctly for empty value [without colors]', false],
    ])('%s', (_, colorize) => {
      const error = new EnumValidationError(
        {
          keyword: 'enum',
          dataPath: '',
          schemaPath: '#/enum',
          params: {
            allowedValues: ['foo', 'bar'],
          },
          message: 'should be equal to one of the allowed values',
        },
        { colorize, data, schema, jsonRaw, jsonAst }
      );

      expect(error.print(schema, '')).toMatchSnapshot();
    });
  });
});
