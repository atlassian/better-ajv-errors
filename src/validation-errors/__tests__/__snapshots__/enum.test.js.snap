// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enum when value is a primitive prints correctly for empty value [without colors] 1`] = `
Array [
  "ENUM should be equal to one of the allowed values",
  "(foo, bar)
",
  "> 1 | \\"baz\\"
    | ^^^^^ 👈🏽  Did you mean bar here?",
]
`;

exports[`Enum when value is a primitive prints correctly for empty value 1`] = `
Array [
  "[31m[1mENUM[22m[39m[31m should be equal to one of the allowed values[39m",
  "[31m(foo, bar)[39m
",
  "[0m[31m[1m>[22m[39m[90m 1 |[39m [32m\\"baz\\"[39m[0m
[0m [90m   |[39m [31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m [31m[1m👈🏽  Did you mean [95mbar[31m here?[22m[39m[0m",
]
`;

exports[`Enum when value is a primitive prints correctly for enum prop [without colors] 1`] = `
Array [
  "ENUM should be equal to one of the allowed values",
  "(foo, bar)
",
  "> 1 | \\"baz\\"
    | ^^^^^ 👈🏽  Did you mean bar here?",
]
`;

exports[`Enum when value is a primitive prints correctly for enum prop 1`] = `
Array [
  "[31m[1mENUM[22m[39m[31m should be equal to one of the allowed values[39m",
  "[31m(foo, bar)[39m
",
  "[0m[31m[1m>[22m[39m[90m 1 |[39m [32m\\"baz\\"[39m[0m
[0m [90m   |[39m [31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m [31m[1m👈🏽  Did you mean [95mbar[31m here?[22m[39m[0m",
]
`;

exports[`Enum when value is a primitive prints correctly for no levenshtein match [without colors] 1`] = `
Array [
  "ENUM should be equal to one of the allowed values",
  "(one, two)
",
  "> 1 | \\"baz\\"
    | ^^^^^ 👈🏽  Unexpected value, should be equal to one of the allowed values",
]
`;

exports[`Enum when value is a primitive prints correctly for no levenshtein match 1`] = `
Array [
  "[31m[1mENUM[22m[39m[31m should be equal to one of the allowed values[39m",
  "[31m(one, two)[39m
",
  "[0m[31m[1m>[22m[39m[90m 1 |[39m [32m\\"baz\\"[39m[0m
[0m [90m   |[39m [31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m [31m[1m👈🏽  Unexpected value, should be equal to one of the allowed values[22m[39m[0m",
]
`;

exports[`Enum when value is an object prints correctly for empty value [without colors] 1`] = `
Array [
  "ENUM should be equal to one of the allowed values",
  "(foo, bar)
",
  "  1 | {
> 2 |   \\"id\\": \\"baz\\"
    |         ^^^^^ 👈🏽  Did you mean bar here?
  3 | }",
]
`;

exports[`Enum when value is an object prints correctly for empty value 1`] = `
Array [
  "[31m[1mENUM[22m[39m[31m should be equal to one of the allowed values[39m",
  "[31m(foo, bar)[39m
",
  "[0m [90m 1 |[39m {[0m
[0m[31m[1m>[22m[39m[90m 2 |[39m   [32m\\"id\\"[39m[33m:[39m [32m\\"baz\\"[39m[0m
[0m [90m   |[39m         [31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m [31m[1m👈🏽  Did you mean [95mbar[31m here?[22m[39m[0m
[0m [90m 3 |[39m }[0m",
]
`;

exports[`Enum when value is an object prints correctly for enum prop [without colors] 1`] = `
Array [
  "ENUM should be equal to one of the allowed values",
  "(foo, bar)
",
  "  1 | {
> 2 |   \\"id\\": \\"baz\\"
    |         ^^^^^ 👈🏽  Did you mean bar here?
  3 | }",
]
`;

exports[`Enum when value is an object prints correctly for enum prop 1`] = `
Array [
  "[31m[1mENUM[22m[39m[31m should be equal to one of the allowed values[39m",
  "[31m(foo, bar)[39m
",
  "[0m [90m 1 |[39m {[0m
[0m[31m[1m>[22m[39m[90m 2 |[39m   [32m\\"id\\"[39m[33m:[39m [32m\\"baz\\"[39m[0m
[0m [90m   |[39m         [31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m [31m[1m👈🏽  Did you mean [95mbar[31m here?[22m[39m[0m
[0m [90m 3 |[39m }[0m",
]
`;

exports[`Enum when value is an object prints correctly for no levenshtein match [without colors] 1`] = `
Array [
  "ENUM should be equal to one of the allowed values",
  "(one, two)
",
  "  1 | {
> 2 |   \\"id\\": \\"baz\\"
    |         ^^^^^ 👈🏽  Unexpected value, should be equal to one of the allowed values
  3 | }",
]
`;

exports[`Enum when value is an object prints correctly for no levenshtein match 1`] = `
Array [
  "[31m[1mENUM[22m[39m[31m should be equal to one of the allowed values[39m",
  "[31m(one, two)[39m
",
  "[0m [90m 1 |[39m {[0m
[0m[31m[1m>[22m[39m[90m 2 |[39m   [32m\\"id\\"[39m[33m:[39m [32m\\"baz\\"[39m[0m
[0m [90m   |[39m         [31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m [31m[1m👈🏽  Unexpected value, should be equal to one of the allowed values[22m[39m[0m
[0m [90m 3 |[39m }[0m",
]
`;
